### PR TITLE
Fix F# properties with arguments resolution in C#

### PR DESCRIPTION
**Pull request by DedSec256**
https://youtrack.jetbrains.com/issue/RIDER-49786

### CODE REVIEW

File: ReSharper.FSharp/src/FSharp.Common/src/Util/Util.fs [hunk 1]
```diff
@@ -153,6 +155,12 @@ module rec FcsUtil =
     let inline (|IdentRange|) (id: Ident) = id.idRange
     let inline (|TypeRange|) (typ: SynType) = typ.Range
 
+    [<Extension; CompiledName "GetAccessRights">]
+    let GetAccessRights(m: FSharpMemberOrFunctionOrValue) =
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Common/src/Util/Util.fs:159]: `CompiledName` seems to add nothing here.

File: ReSharper.FSharp/src/FSharp.Psi/src/Impl/DeclaredElement/FSharpPropertyAccessor.cs [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+using JetBrains.ReSharper.Psi.Impl;
+using JetBrains.ReSharper.Psi.Modules;
+using JetBrains.ReSharper.Psi.Resolve;
+using JetBrains.ReSharper.Psi.Tree;
+using JetBrains.Util;
+using JetBrains.Util.DataStructures;
+
+namespace JetBrains.ReSharper.Plugins.FSharp.Psi.Impl.DeclaredElement
+{
+  public class FSharpPropertyAccessor : IAccessor
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi/src/Impl/DeclaredElement/FSharpPropertyAccessor.cs:22]: This class seems to reimplement too many things. Why is it better than reusing the available base ones?

File: ReSharper.FSharp/src/FSharp.Psi/src/Impl/DeclaredElement/FSharpPropertyBase.cs [hunk 3]
```diff
@@ -18,17 +17,23 @@ internal abstract class FSharpPropertyMemberBase<TDeclaration> : FSharpPropertyB
...
 
-      IsReadable = prop.HasGetterMethod || prop.IsPropertyGetterMethod ||
-                   prop.IsModuleValueOrMember && !prop.IsMember;
+      var isReadable = IsReadable = prop.HasGetterMethod || prop.IsPropertyGetterMethod ||
+                                    prop.IsModuleValueOrMember && !prop.IsMember;
 
-      IsWritable = prop.IsMutable || prop.HasSetterMethod || prop.IsPropertySetterMethod;
+      var isWritable = IsWritable = prop.IsMutable || prop.HasSetterMethod || prop.IsPropertySetterMethod;
+
+      Getter = isReadable ? new FSharpPropertyAccessor(prop.GetterMethod, this) : null;
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi/src/Impl/DeclaredElement/FSharpPropertyBase.cs:25]: Do we really need to store it in all cases? Wouldn't the implicit one be sufficient in the most cases?

File: ReSharper.FSharp/src/FSharp.Psi/src/Impl/Tree/MemberDeclaration.cs [hunk 4]
```diff
@@ -53,9 +53,9 @@ protected override IDeclaredElement CreateDeclaredElement(FSharpSymbol fcsSymbol
       if (property != null)
       {
         var cliEvent = property.EventForFSharpProperty?.Value;
-        return cliEvent != null
-          ? (ITypeMember) new FSharpCliEvent<MemberDeclaration>(this)
-          : new FSharpProperty<MemberDeclaration>(this, property);
+        if (cliEvent != null) return new FSharpCliEvent<MemberDeclaration>(this);
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi/src/Impl/Tree/MemberDeclaration.cs:56]: Why this change is in this PR?

File: ReSharper.FSharp/src/FSharp.Psi/src/Impl/Cache2/Parts/FSharpTypeMembersOwnerTypePart.cs [hunk 5]
```diff
@@ -96,5 +98,29 @@ public override IEnumerable<ITypeElement> GetSuperTypeElements()
...
+          if (prop.Getter != null) result.Add(prop.Getter);
+          if (prop.Setter != null) result.Add(prop.Setter);
+        }
+        else
+        {
+          result.Add(declaredElement);
+        }
+      }
+
+      return result.ResultingList();
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi/src/Impl/Cache2/Parts/FSharpTypeMembersOwnerTypePart.cs:123]: This method duplicates the base one almost completely.

File: ReSharper.FSharp/src/FSharp.Psi/src/FSharp.psi [hunk 6]
```diff
@@ -1310,6 +1310,7 @@ accessorsNamesClause:
   );
 
 accessorDeclaration:
+  accessModifier{ACCESS_MODIFIER, AccessModifier}?
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi/src/FSharp.psi:1313]: Is there any test case covering the accessor modifiers?

**Review by auduchinok — CHANGES_REQUESTED** [review]
Could you please try to reduce added duplications?
Also, tests are missing, please add them.

File: ReSharper.FSharp/src/FSharp.Psi/src/Impl/Cache2/Parts/FSharpTypeMembersOwnerTypePart.cs (hunk 5 above)
Comment (DedSec256) [reply ReSharper.FSharp/src/FSharp.Psi/src/Impl/Cache2/Parts/FSharpTypeMembersOwnerTypePart.cs:106]: not anymore (:

File: ReSharper.FSharp/src/FSharp.Psi/src/Impl/DeclaredElement/CompilerGenerated/FSharpGeneratedPropertyAccessor.cs [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+using JetBrains.ReSharper.Plugins.FSharp.Psi.Impl.Pointers;
+using JetBrains.ReSharper.Psi;
+using JetBrains.ReSharper.Psi.Pointers;
+
+namespace JetBrains.ReSharper.Plugins.FSharp.Psi.Impl.DeclaredElement.CompilerGenerated
+{
+  internal class FSharpGeneratedPropertyAccessor : FSharpGeneratedMethodBase, IFSharpGeneratedFromProperty
+  {
+    private readonly bool myIsGetter;
+    private readonly IAccessor myAccessor;
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi/src/Impl/DeclaredElement/CompilerGenerated/FSharpGeneratedPropertyAccessor.cs:12]: Why does an accessor instance hold a reference to another accessor?

File: ReSharper.FSharp/src/FSharp.Psi/src/Impl/DeclaredElement/FSharpPropertyAccessor.cs [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+using JetBrains.ReSharper.Psi.ExtensionsAPI;
+using JetBrains.ReSharper.Psi.Modules;
+using JetBrains.ReSharper.Psi.Resolve;
+using JetBrains.ReSharper.Psi.Tree;
+using JetBrains.Util;
+using JetBrains.Util.DataStructures;
+
+namespace JetBrains.ReSharper.Plugins.FSharp.Psi.Impl.DeclaredElement
+{
+  public class FSharpPropertyAccessor : IAccessor, IFSharpTypeParametersOwner
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi/src/Impl/DeclaredElement/FSharpPropertyAccessor.cs:18]: This type seems to duplicate `ImplicitAccessor` almost completely.

File: ReSharper.FSharp/src/FSharp.Psi.Features/src/Search/FSharpGeneratedElementsSearch.fs [hunk 9]
```diff
@@ -16,7 +16,8 @@ type FSharpGeneratedElementsSearch() =
     override x.IsContextApplicable(dataContext: IDataContext) =
         match dataContext.GetData(PsiDataConstants.DECLARED_ELEMENTS) with
         | null -> false
-        | elements -> elements |> Seq.forall (fun el -> el :? IFSharpGeneratedFromUnionCase)
+        | elements -> elements |> Seq.forall (fun el -> el :? IFSharpGeneratedFromUnionCase ||
+                                                        el :? IFSharpGeneratedFromProperty )
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi.Features/src/Search/FSharpGeneratedElementsSearch.fs:20]: Please move piped expressions to new lines instead.

File: ReSharper.FSharp/src/FSharp.Psi/src/IFSharpTypeMember.cs [hunk 10]
```diff
@@ -39,4 +39,8 @@ public interface IFSharpGeneratedFromOtherElement : IFSharpGeneratedElement
   public interface IFSharpGeneratedFromUnionCase : IFSharpGeneratedFromOtherElement
   {
   }
+
+  public interface IFSharpGeneratedFromProperty : IFSharpGeneratedFromOtherElement
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi/src/IFSharpTypeMember.cs:43]: Was this interface introduced for a single implementation only? What other types you expect there?
Would it be better to add an interface like `IFSharpGetterInstead`?

File: ReSharper.FSharp/src/FSharp.Psi/src/Impl/DeclaredElement/CompilerGenerated/FSharpGeneratedPropertyAccessor.cs (hunk 7 above)
Comment (DedSec256) [reply ReSharper.FSharp/src/FSharp.Psi/src/Impl/DeclaredElement/CompilerGenerated/FSharpGeneratedPropertyAccessor.cs:12]: Just for proof of concept. I'm going to make the appropriate changes in the platform.

File: ReSharper.FSharp/src/FSharp.Psi/src/IFSharpTypeMember.cs (hunk 10 above)
Comment (DedSec256) [reply ReSharper.FSharp/src/FSharp.Psi/src/IFSharpTypeMember.cs:43]: Another implementation that is not in this PR is the property-indexer

Comment (DedSec256) [reply ReSharper.FSharp/src/FSharp.Psi/src/IFSharpTypeMember.cs:43]: What about ```IFSharpGeneratedAccessor```?

Comment (auduchinok) [reply ReSharper.FSharp/src/FSharp.Psi/src/IFSharpTypeMember.cs:43]: Right, it'd be better.

File: ReSharper.FSharp/test/data/features/refactorings/rename/Generated members - Property accessors 01.cs [hunk 11]
```diff
@@ -0,0 +1,13 @@
+//${NEW_NAME:Zzz}
```
Comment (auduchinok) [inline ReSharper.FSharp/test/data/features/refactorings/rename/Generated members - Property accessors 01.cs:1]: Any reason the default name wouldn't work?

File: ReSharper.FSharp/src/FSharp.Psi/src/Impl/Cache2/Parts/FSharpTypeMembersOwnerTypePart.cs (hunk 5 above)
Comment (auduchinok) [reply ReSharper.FSharp/src/FSharp.Psi/src/Impl/Cache2/Parts/FSharpTypeMembersOwnerTypePart.cs:103]: Yes, now it should probably belong to `FSharpTypeElementDeclarationBase.ProcessChildren`. It should also take declared elements from accessors declarations, not from the property declared element.

File: ReSharper.FSharp/src/FSharp.Psi/src/Impl/GeneratedMembersUtil.cs [hunk 12]
```diff
@@ -104,5 +105,27 @@ public static IEnumerable<IDeclaredElement> GetGeneratedMembers([NotNull] this I
 
       return result;
     }
+
+    public static IList<ITypeMember> GetGeneratedMembers(this IFSharpProperty property)
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi/src/Impl/GeneratedMembersUtil.cs:109]: This name seems wrong. It only fetches declared accessors, nothing that is additionally generated.

**Review by auduchinok — COMMENTED** [review]
I think binding explicitly declared accessor declared elements to actual declarations would be a better approach, since now you produce the elements without any connection to their declarations and it doesn't seem right.

File: ReSharper.FSharp/src/FSharp.Psi/src/IFSharpTypeMember.cs [hunk 13]
```diff
@@ -37,4 +37,9 @@ public interface IFSharpGeneratedFromOtherElement : IFSharpGeneratedElement
   public interface IFSharpGeneratedFromUnionCase : IFSharpGeneratedFromOtherElement
   {
   }
+
+  public interface IFSharpGeneratedAccessor : IFSharpGeneratedFromOtherElement
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi/src/IFSharpTypeMember.cs:41]: It's not generated, it's explicitly declared.

File: ReSharper.FSharp/src/FSharp.Psi/src/FSharp.psi [hunk 14]
```diff
@@ -1120,6 +1120,8 @@ interface fSharpTypeMemberDeclaration:
   letBindingsDeclaration |
   overridableMemberDeclaration |
   doStatement |
+  accessorDeclaration |
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi/src/FSharp.psi:1123]: Why is it needed?

File: ReSharper.FSharp/src/FSharp.Psi/src/Impl/Cache2/Parts/FSharpTypeMembersOwnerTypePart.cs (hunk 5 above)
Comment (auduchinok) [reply ReSharper.FSharp/src/FSharp.Psi/src/Impl/Cache2/Parts/FSharpTypeMembersOwnerTypePart.cs:103]: Still not resolved.

File: ReSharper.FSharp/src/FSharp.Psi/src/Impl/DeclaredElement/CompilerGenerated/FSharpPropertyAccessorMethod.cs [hunk 15]
```diff
@@ -1,12 +1,15 @@
...
 using JetBrains.ReSharper.Plugins.FSharp.Psi.Impl.Cache2.Parts;
 using JetBrains.ReSharper.Plugins.FSharp.Psi.Impl.Pointers;
+using JetBrains.ReSharper.Plugins.FSharp.Psi.Tree;
 using JetBrains.ReSharper.Psi;
 using JetBrains.ReSharper.Psi.Pointers;
 
 namespace JetBrains.ReSharper.Plugins.FSharp.Psi.Impl.DeclaredElement.CompilerGenerated
 {
-  internal class FSharpPropertyAccessorMethod : FSharpGeneratedMethodBase, IFSharpGeneratedAccessor
+  internal class FSharpPropertyAccessorMethod : FSharpGeneratedMethodBase, IFSharpGeneratedAccessor, IFSharpMember
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi/src/Impl/DeclaredElement/CompilerGenerated/FSharpPropertyAccessorMethod.cs:12]: Why `Method` is needed? Accessor is always a method, isn't it?

File: ReSharper.FSharp/src/FSharp.Psi/src/Impl/DeclaredElement/CompilerGenerated/FSharpPropertyAccessorMethod.cs [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+using JetBrains.ReSharper.Plugins.FSharp.Psi.Impl.Pointers;
+using JetBrains.ReSharper.Plugins.FSharp.Psi.Tree;
+using JetBrains.ReSharper.Psi;
+using JetBrains.ReSharper.Psi.Pointers;
+
+namespace JetBrains.ReSharper.Plugins.FSharp.Psi.Impl.DeclaredElement.CompilerGenerated
+{
+  internal class FSharpPropertyAccessorMethod : FSharpGeneratedMethodBase, IFSharpGeneratedAccessor, IFSharpMember
+  {
+    private readonly IAccessor myAccessor;
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi/src/Impl/DeclaredElement/CompilerGenerated/FSharpPropertyAccessorMethod.cs:14]: Why it's needed?

File: ReSharper.FSharp/src/FSharp.Psi/src/Impl/DeclaredElement/FSharpPropertyWithExplicitAccessors.cs [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+      mySetter = accessors.TryGet(AccessorKind.SETTER);
+    }
+
+    public override bool IsReadable => myGetter != null;
+    public override bool IsWritable => mySetter != null;
+    public override IAccessor Getter => IsReadable ? myGetter.DeclaredElement as IAccessor : null;
+    public override IAccessor Setter => IsWritable ? mySetter.DeclaredElement as IAccessor : null;
+    public override AccessRights GetAccessRights() => AccessRights.PRIVATE;
+    public AccessRights RepresentationAccessRights => base.GetAccessRights();
+    public override IList<IParameter> Parameters => this.GetParameters(Mfv);
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi/src/Impl/DeclaredElement/FSharpPropertyWithExplicitAccessors.cs:27]: Shouldn't these parameters belong to accessor?

File: ReSharper.FSharp/src/FSharp.Psi/src/Impl/DeclaredElement/CompilerGenerated/FSharpPropertyAccessorMethod.cs (hunk 15 above)
Comment (DedSec256) [reply ReSharper.FSharp/src/FSharp.Psi/src/Impl/DeclaredElement/CompilerGenerated/FSharpPropertyAccessorMethod.cs:12]: Resolve in C# explicitly separates accessors from methods

Comment (auduchinok) [reply ReSharper.FSharp/src/FSharp.Psi/src/Impl/DeclaredElement/CompilerGenerated/FSharpPropertyAccessorMethod.cs:12]: > Resolve in C# explicitly separates accessors from methods.

What do you mean? `IAccessor` inherits `IMethod`.

File: ReSharper.FSharp/src/FSharp.Psi/src/FSharp.psi (hunk 14 above)
Comment (DedSec256) [reply ReSharper.FSharp/src/FSharp.Psi/src/FSharp.psi:1123]: At some point I needed it, but now I don't need it, I will delete

File: ReSharper.FSharp/src/FSharp.Psi/src/Impl/DeclaredElement/CompilerGenerated/FSharpPropertyAccessorMethod.cs (hunk 15 above)
Comment (auduchinok) [reply ReSharper.FSharp/src/FSharp.Psi/src/Impl/DeclaredElement/CompilerGenerated/FSharpPropertyAccessorMethod.cs:12]: OK, this accessor is not an accessor, I see.

File: ReSharper.FSharp/src/FSharp.Psi/src/Impl/DeclaredElement/FSharpPropertyWithExplicitAccessors.cs (hunk 17 above)
Comment (DedSec256) [reply ReSharper.FSharp/src/FSharp.Psi/src/Impl/DeclaredElement/FSharpPropertyWithExplicitAccessors.cs:27]: This is necessary in order for ReSharper to correctly display tooltips for explicit accessors in the future. When ReSharper reads such properties from the physical assembly, they are presented there with their arguments in a similar way.

**Comment by auduchinok** [comment]
I've pushed small cleanup and fixup. @DedSec256 Could you fix the test names, please? Most of them say accessors are generated which is wrong.

**Review by auduchinok — APPROVED** [review]
@DedSec256 Thanks! It looks very good now.

**Pull request merged**